File: forgepay/packages/sdk-js/src/resources/customers.ts
import type { FPHttpClient } from '../client.js';
import type { Customer, CustomerCreateParams, ListResponse } from '../types.js';

export class CustomersResource {
  constructor(private readonly client: FPHttpClient) {}

  /**
   * Create a customer.
   *
   * ```ts
   * const customer = await forgepay.customers.create({
   *   email: 'alice@acme.io',
   *   name: 'Alice Smith',
   * });
   * ```
   */
  async create(params: CustomerCreateParams): Promise<Customer> {
    return this.client.post<Customer>('/v1/customers', params);
  }

  async retrieve(customerId: string): Promise<Customer> {
    return this.client.get<Customer>(`/v1/customers/${customerId}`);
  }

  async update(customerId: string, params: Partial<CustomerCreateParams>): Promise<Customer> {
    return this.client.patch<Customer>(`/v1/customers/${customerId}`, params);
  }

  async list(params?: { limit?: number; cursor?: string }): Promise<ListResponse<Customer>> {
    return this.client.get<ListResponse<Customer>>('/v1/customers', params as Record<string, string | number | undefined>);
  }

  async delete(customerId: string): Promise<{ id: string; deleted: true }> {
    return this.client.delete(`/v1/customers/${customerId}`);
  }
}

File: forgepay/packages/sdk-python/forgepay/resources/customers.py
from __future__ import annotations

from typing import Any

from forgepay._http import SyncTransport, AsyncTransport
from forgepay.types import Customer, CustomerCreateParams, ListResponse


class CustomersResource:
    def __init__(self, transport: SyncTransport) -> None:
        self._t = transport

    def create(self, **kwargs: Any) -> Customer:
        params = CustomerCreateParams(**kwargs)
        raw    = self._t.request("POST", "/v1/customers", json=params.model_dump(exclude_none=True))
        return Customer.model_validate(raw)

    def retrieve(self, customer_id: str) -> Customer:
        return Customer.model_validate(self._t.request("GET", f"/v1/customers/{customer_id}"))

    def update(self, customer_id: str, **kwargs: Any) -> Customer:
        raw = self._t.request("POST", f"/v1/customers/{customer_id}", json=kwargs)
        return Customer.model_validate(raw)

    def list(self, **params: Any) -> ListResponse[Customer]:  # type: ignore[type-arg]
        raw = self._t.request("GET", "/v1/customers", params=params or None)
        return ListResponse[Customer].model_validate(raw)

    def delete(self, customer_id: str) -> dict[str, Any]:
        return self._t.request("DELETE", f"/v1/customers/{customer_id}")  # type: ignore[return-value]


class AsyncCustomersResource:
    def __init__(self, transport: AsyncTransport) -> None:
        self._t = transport

    async def create(self, **kwargs: Any) -> Customer:
        params = CustomerCreateParams(**kwargs)
        raw    = await self._t.request(
            "POST", "/v1/customers", json=params.model_dump(exclude_none=True)
        )
        return Customer.model_validate(raw)

    async def retrieve(self, customer_id: str) -> Customer:
        return Customer.model_validate(
            await self._t.request("GET", f"/v1/customers/{customer_id}")
        )

    async def update(self, customer_id: str, **kwargs: Any) -> Customer:
        raw = await self._t.request("POST", f"/v1/customers/{customer_id}", json=kwargs)
        return Customer.model_validate(raw)

    async def list(self, **params: Any) -> ListResponse[Customer]:  # type: ignore[type-arg]
        raw = await self._t.request("GET", "/v1/customers", params=params or None)
        return ListResponse[Customer].model_validate(raw)

    async def delete(self, customer_id: str) -> dict[str, Any]:
        return await self._t.request("DELETE", f"/v1/customers/{customer_id}")  # type: ignore[return-value]
